fix(aucmd_win): ensure aucmd_win stays floating

seandewar opened the pull request:
Nvim uses a floating window for the autocmd window, but in certain situations,
it can be made non-floating (`:wincmd J`), which can cause issues due to the
previous setup and cleanup logic for a non-floating aucmd_win being removed from
aucmd_prepbuf and aucmd_restbuf.

This can cause glitchiness and crashes due to the aucmd_win's frame being
invalid after closing its tabpage, for example.

Ensure aucmd_win cannot be made non-floating. The only place this happens is in
win_split_ins.

Closes #13699.
Also fixes #13104 after re-applying the reverted patches in #13123.

seandewar committed:
fix(aucmd_win): ensure aucmd_win stays floating

Nvim uses a floating window for the autocmd window, but in certain situations,
it can be made non-floating (`:wincmd J`), which can cause issues due to the
previous setup and cleanup logic for a non-floating aucmd_win being removed from
aucmd_prepbuf and aucmd_restbuf.

This can cause glitchiness and crashes due to the aucmd_win's frame being
invalid after closing its tabpage, for example.

Ensure aucmd_win cannot be made non-floating. The only place this happens is in
win_split_ins if new_wp != NULL.

bfredl commented:
Hmm, I would go one step further and forbid it to be made non-floating in the first place? any downsides to that (like any popular plugin we would upset)?

seandewar commented:
No idea. I can't see any legitimate use for moving around the aucmd window that isn't a mistake (like in #13699 where people expected it to move the qf window again).

I wonder if it's worth raising an error with that approach (or just silently do nothing?)

bfredl commented:
yes, perhaps the simplest is to just  make any `wincmd J` like command silently do nothing.

seandewar commented:
Hmm, at the very least I've gone ahead and made splitting return `FAIL` from `win_split_ins`. Calling code can technically complain about that if they want to, but the window commands and most other places are fine with it.

I still need to recompute the window layout in `aucmd_restbuf` as in some cases the return value of `win_split_ins` is ignored and calling code does some resizing anyway.

bfredl commented on src/nvim/autocmd.c:
```diff
@@ -1152,6 +1152,7 @@ void aucmd_prepbuf(aco_save_T *aco, buf_T *buf)
     globaldir = NULL;
 
     block_autocmds();  // We don't want BufEnter/WinEnter autocommands.
+    make_snapshot(SNAP_AUCMD_IDX);
```
Is this really needed, now what we banned inserting the aucmdwin? saving/restoring window layout was the anti-pattern we aimed at getting rid of in the first place..

seandewar replied:
Tests fail without it. If I remember correctly, the problem is that even if we do not allow `aucmd_win` to be messed around in `win_split_ins`, calling code still ignores the return value and messes with the window layout?

bfredl replied:
So, shouldn't we fix the calling code then?

seandewar replied:
Looks like the other way around, layout is messed with _before_ `win_split_ins`, in preparation for it.

This only seems to be called in places that want to move a window to a different split.
To do this, `win_split_ins` is called with `new_wp != NULL`, to insert an existing window into a new split.
Layout is messed with to remove the old split (otherwise you'd have 2 splits referring to the same window...).

Calling `win_split_ins` first to check its return value beforehand would potentially cause problems with other windows, I think (e.g: you may get `e_noroom` because the old split is still taking up space).

Instead I just test for `aucmd_win` before the only 2 places that mess with the layout.
`win_split` isn't affected as it creates a new split of a new window from `curwin` (`new_wp == NULL`; `aucmd_win` should be allowed to do that).

seandewar committed:
refactor(aucmd_win): remove need to restore window layout

There are some places that mess with the window layout in preparation for moving
a window to a different split (win_split_ins called with new_wp != NULL).

This means the window layout can change slightly even if win_split_ins fails.

This is why it was still needed to restore the window layout in
aucmd_{prep,rest}buf even if we disallow win_split_ins from making aucmd_win
non-floating by moving it into a split.

We can just skip messing with the layout in such places if we're dealing with
the aucmd_win.

zeertzjq commented:
BTW, I noticed that Vim makes autocmd window the first window in the window list (having window number 1), whereas Nvim makes it the last one since #9798. I'm not sure which one is better, but I noticed this when trying to port <https://github.com/vim/vim/commit/832adf9bb8cd39d8e982d8a35ed8a6d39b974494> to Nvim.

bfredl commented:
I'd say let keep the behavior of being last in the window list, and update ported tests as needed. a plugin which breaks on that could just as well break on aucmd_win being floating at all.
